velox/connectors/hive/iceberg/PositionalDeleteFileReader.cpp

```diff
@@ -0,0 +1,261 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/iceberg/PositionalDeleteFileReader.h"
+
+#include "velox/connectors/hive/HiveConnectorUtil.h"
+#include "velox/connectors/hive/TableHandle.h"
+#include "velox/connectors/hive/iceberg/IcebergDeleteFile.h"
+#include "velox/connectors/hive/iceberg/IcebergMetadataColumns.h"
+#include "velox/dwio/common/ReaderFactory.h"
+
+namespace facebook::velox::connector::hive::iceberg {
+
+PositionalDeleteFileReader::PositionalDeleteFileReader(
+    const IcebergDeleteFile& deleteFile,
+    const std::string& baseFilePath,
+    FileHandleFactory* fileHandleFactory,
+    const ConnectorQueryCtx* connectorQueryCtx,
+    folly::Executor* executor,
+    const std::shared_ptr<HiveConfig> hiveConfig,
+    std::shared_ptr<io::IoStatistics> ioStats,
+    dwio::common::RuntimeStatistics& runtimeStats,
+    uint64_t splitOffset,
+    const std::string& connectorId)
+    : deleteFile_(deleteFile),
+      baseFilePath_(baseFilePath),
+      fileHandleFactory_(fileHandleFactory),
+      executor_(executor),
+      connectorQueryCtx_(connectorQueryCtx),
+      hiveConfig_(hiveConfig),
+      ioStats_(ioStats),
+      pool_(connectorQueryCtx->memoryPool()),
+      filePathColumn_(IcebergMetadataColumn::icebergDeleteFilePathColumn()),
+      posColumn_(IcebergMetadataColumn::icebergDeletePosColumn()),
+      splitOffset_(splitOffset),
+      deleteSplit_(nullptr),
+      deleteRowReader_(nullptr),
+      deletePositionsOutput_(nullptr),
+      deletePositionsOffset_(0),
+      endOfFile_(false) {
+  VELOX_CHECK_EQ(deleteFile_.content, FileContent::kPositionalDeletes);
+
+  if (deleteFile_.recordCount == 0) {
+    return;
+  }
+
+  // TODO: check if the lowerbounds and upperbounds in deleteFile overlap with
+  //  this batch. If not, no need to proceed.
+
+  auto outputRowType = ROW({posColumn_->name}, {posColumn_->type});
+  common::Subfield posSubfield(posColumn_->name);
+  folly::F14FastMap<std::string, std::vector<const common::Subfield*>>
+      outputSubfieldsMap;
+  outputSubfieldsMap[posColumn_->name].push_back(&posSubfield);
+
+  std::vector<std::string> deleteColumnNames(
+      {filePathColumn_->name, posColumn_->name});
+  std::vector<std::shared_ptr<const Type>> deleteColumnTypes(
+      {filePathColumn_->type, posColumn_->type});
+  RowTypePtr deleteFileSchema =
+      ROW(std::move(deleteColumnNames), std::move(deleteColumnTypes));
+
+  auto scanSpec = makeScanSpec(
```

Comment:
You don't need `makeScanSpec` here, just create a scan spec with 2 columns and add a filter on path

Reply:
@Yuhta I think using makeScanSpec() is better here. To manually write the code
```
  std::vector<std::string> values = {baseFilePath_};
  std::unique_ptr<common::Filter> pathFilter =
      std::make_unique<common::BytesValues>(values, false);
  auto posRange = ...;  // To be added
  std::unique_ptr<common::Filter> posFilter =
      std::make_unique<common::BigintRange>(posRange, false);

  auto scanSpec = std::make_shared<common::ScanSpec>("<root>");
  scanSpec->addField(posColumn_->name, 0);

  auto* pathSpec =
      scanSpec->getOrCreateChild(common::Subfield(filePathColumn_->name));
  pathSpec->setFilter(std::move(pathFilter));
  auto* posSpec = scanSpec->childByName(posColumn_->name);
  posSpec->setFilter(std::move(posFilter));
```
This is much longer and harder to read than just 
```
      auto scanSpec = makeScanSpec(
          outputRowType,
          outputSubfieldsMap,
          createFilters(),
          deleteFileSchema,
          {},
          pool_);
```
It's better to confine the creation of the scanspec into a central location, i.e. in makeScanSpec(). It's more modular and its interface is relatively clearer. Users/Readers don't have to understand what is done inside, but just by providing the required row type, filters and subfields, a ScanSpec can be made. (btw. I think makeScanSpec() can be potentially simplified a bit but that's not crucial now)

Moreover, the equality delete file reader will also need to make ScanSpec. Why do we repeat the code that's already there? Also the schema of the equality delete file is not known until query run time.

If you worry about the cost to create the ScanSpec using makeScanSpec(), it shouldn't be a concern here IMHO. This operation is once per split, and even in makeScanSpec(), the operations done inside should not be fundamentally a lot more than the manual creation, because the positional delete files can have at most 2 or 3 columns. But if our future microbenchmark shows the ScanSpec creation actually affects the performance, I'll surely change it myself.

Reply:
By the way, I tried your suggestion https://github.com/facebookincubator/velox/pull/7362#discussion_r1412317067 but it doesn't work because the channel is not set up correctly. 
```
auto spec = std::make_shared<common::ScanSpec>("<root>");
    spec->addAllChildFields(*rowType);   // This would set the channel 0 and 1 but the path spce should be -1
    auto* pathSpec = spec->childByName("path");
    pathSpec->setFilter(std::make_unique<common::BytesValues>(std::vector<std::string>{path}, false));
    pathSpec->setProjectOut(false);
```
You see, manually creating the ScanSpec is not as straightforward as you thought. Even if we set it right at this moment, future modifications might make it wrong easily. Since there is already a clear API makeScanSpec(), I think it's better to reuse it.

Reply:
There are 2 ways you can solve the channel.  What I did is to have the result batch having 2 columns, but since the path is not projected out, it is not really used.  And the result batch is reused so the waste is not large.  Another way is just reset the channel on path.  Also the comparison is not fair, the real comparison is

```c++
  SubfieldFilters subfieldFilters;
  std::vector<std::string> values = {baseFilePath_};
  std::unique_ptr<common::Filter> pathFilter =
      std::make_unique<common::BytesValues>(values, false);
  subfieldFilters[common::Subfield(filePathColumn_->name)] =
      std::move(pathFilter);

  common::Subfield posSubfield(posColumn_->name);
  folly::F14FastMap<std::string, std::vector<const common::Subfield*>>
      outputSubfieldsMap;
  outputSubfieldsMap[posColumn_->name].push_back(&posSubfield);

  auto scanSpec = makeScanSpec(
          outputRowType,
          outputSubfieldsMap,
          createFilters(),
          deleteFileSchema,
          {},
          pool_);
```

vs

```c++
  std::vector<std::string> values = {baseFilePath_};
  std::unique_ptr<common::Filter> pathFilter =
      std::make_unique<common::BytesValues>(values, false);
  auto scanSpec = std::make_shared<common::ScanSpec>("<root>");
  scanSpec->addField(posColumn_->name, 0);
  auto* pathSpec =
      scanSpec->getOrCreateChild(common::Subfield(filePathColumn_->name));
  pathSpec->setFilter(std::move(pathFilter));
```

Reply:
> There are 2 ways you can solve the channel. What I did is to have the result batch having 2 columns, but since the path is not projected out, it is not really used. And the result batch is reused so the waste is not large. Another way is just reset the channel on path. Also the comparison is not fair, the real comparison is
> 
> ```c++
>   SubfieldFilters subfieldFilters;
>   std::vector<std::string> values = {baseFilePath_};
>   std::unique_ptr<common::Filter> pathFilter =
>       std::make_unique<common::BytesValues>(values, false);
>   subfieldFilters[common::Subfield(filePathColumn_->name)] =
>       std::move(pathFilter);
> 
>   common::Subfield posSubfield(posColumn_->name);
>   folly::F14FastMap<std::string, std::vector<const common::Subfield*>>
>       outputSubfieldsMap;
>   outputSubfieldsMap[posColumn_->name].push_back(&posSubfield);
> 
>   auto scanSpec = makeScanSpec(
>           outputRowType,
>           outputSubfieldsMap,
>           subfieldFilters,
>           deleteFileSchema,
>           {},
>           pool_);
> ```
> 
> vs
> 
> ```c++
>   std::vector<std::string> values = {baseFilePath_};
>   std::unique_ptr<common::Filter> pathFilter =
>       std::make_unique<common::BytesValues>(values, false);
>   auto scanSpec = std::make_shared<common::ScanSpec>("<root>");
>   scanSpec->addField(posColumn_->name, 0);
>   auto* pathSpec =
>       scanSpec->getOrCreateChild(common::Subfield(filePathColumn_->name));
>   pathSpec->setFilter(std::move(pathFilter));
> ```

@Yuhta Thanks for your writeup and suggestion. Yes this worked for me even originally, but it took me a couple of hours to debug, build and test it last week before I decided to switch back to `makeScanSpec()`. Other developer may need to spend similar time if he/she needs to modify it in the future, e.g. adding the filter for the `_pos` column. My real point is that building `ScanSpec` is not as straightforward as you thought, and it's easy to make it wrong, especially for new developers who don't know the internals of `ScanSpec`. He/she will need to read the ScanSpec class implementation to understand what's going on, and their time is not free. IMHO It's better to encapsulate it in one place, which is `makeScanSpec()`, and make it have clean and false-proof interface. As long as a developer passes in the correct arguments, the ScanSpec coming out of it cannot be wrong. Also, having duplicate logic is worse than having slightly more lines IMHO. Suppose you change the logic in `makeScanSpec()` in the future, then this place, and also the future EqualityDeleteFileReader, will need to be updated too.  So it seems to me calling into the same util function is good for both engineering efficiency and code cleanliness, if the performance is not impacted. 

However to make it a little better, I just extracted a `toSubfieldsMap()` function in HiveConnectorUtil.h/cpp, which can also be used by the future EqualityDeleteFileReader. With this change, the code is actually
```
  auto outputRowType = ROW({posColumn_->name}, {posColumn_->type});
  auto scanSpec = makeScanSpec(
      outputRowType,
      toSubfieldsMap(outputRowType),
      createFilters(),
      deleteFileSchema,
      {},
      pool_);
```
versus
```
  std::vector<std::string> values = {baseFilePath_};
  std::unique_ptr<common::Filter> pathFilter =
      std::make_unique<common::BytesValues>(values, false);
  auto scanSpec = std::make_shared<common::ScanSpec>("<root>");
  scanSpec->addField(posColumn_->name, 0);
  auto* pathSpec =
      scanSpec->getOrCreateChild(common::Subfield(filePathColumn_->name));
  pathSpec->setFilter(std::move(pathFilter));
```
The #lines is similar, but I think the first is easier to read and less error-prone.

The example I gave last time also considered some additional filter on the `_pos` column. Similarly the future EqualityDeleteFileReader will have multiple filters and I'll add them step by step. With more filters being added, this part of code will become longer if too the latter approach, and that was indeed as it was shown last time

```
  std::vector<std::string> values = {baseFilePath_};
  std::unique_ptr<common::Filter> pathFilter =
      std::make_unique<common::BytesValues>(values, false);
  auto posRange = ...;  // To be added
  std::unique_ptr<common::Filter> posFilter =
      std::make_unique<common::BigintRange>(posRange, false);
  auto scanSpec = std::make_shared<common::ScanSpec>("<root>");
  scanSpec->addField(posColumn_->name, 0);
  auto* pathSpec =
      scanSpec->getOrCreateChild(common::Subfield(filePathColumn_->name));
  pathSpec->setFilter(std::move(pathFilter));
  auto* posSpec = scanSpec->childByName(posColumn_->name);
  posSpec->setFilter(std::move(posFilter));
```
You may want to put these lines into an internal function, but why not use an existing one if there is no difference in performance?

Speaking of which, if you don't think my concerns are real and still prefer the latter way, I'll update the PR as you suggested. Please let me know. Thanks!